compiler/rustc_mir_transform/src/uninhabited_enum_branching.rs

```diff
@@ -71,6 +72,28 @@ fn variant_discriminants<'tcx>(
     }
 }
 
+/// Ensures that the `otherwise` branch leads to an unreachable bb, returning `None` if so and a new
+/// bb to use as the new target if not.
+fn ensure_otherwise_unreachable<'tcx>(
+    body: &Body<'tcx>,
+    targets: &SwitchTargets,
+) -> Option<BasicBlockData<'tcx>> {
+    let otherwise = targets.otherwise();
+    let bb = &body.basic_blocks()[otherwise];
+    if bb.statements.len() == 0
+        && bb.terminator().kind == TerminatorKind::Unreachable
+        && !bb.is_cleanup
+    {
+        return None;
+    }
+
+    let new_block = BasicBlockData::new(Some(Terminator {
+        source_info: bb.terminator().source_info,
+        kind: TerminatorKind::Unreachable,
+    }));
```

Comment:
The new block should preserve the `is_cleanup` flag, so that normal (non-unwinding) edges only connect blocks with the same value of the flag.

For example this would result in invalid MIR for:

```rust
#![feature(never_type)]

pub enum E {
    A(String, String),
    B(String, !),
}

pub fn f(e: E) {
    match e {
        E::A(_a, ..) => {},
        E::B(_b, ..) => {},
    }
    g();
}

pub fn g() { }
```

Reply:
Thanks for the catch, I've updated this. Question about the semantics here: Is it guaranteed that cleanup and non-cleanup basic blocks are only connected by the `unwind` edges in `call` terminators? Or can they be made to intersect in a more complicated way?

Reply:
Non-cleanup blocks are connected to cleanup blocks only through unwind edges. This is validated by `check_edge` function in `compiler/rustc_const_eval/src/transform/validate.rs`.